core/action.class.inc.php

```diff
@@ -516,26 +527,64 @@ protected function _DoExecute($oTrigger, $aContextArgs, &$oLog)
 				{
 					case EMAIL_SEND_OK:
 						return "Sent";
-	
+
 					case EMAIL_SEND_PENDING:
 						return "Pending";
-	
+
 					case EMAIL_SEND_ERROR:
 						return "Errors: ".implode(', ', $aErrors);
 				}
 			}
-		}
-		else
-		{
-			if (is_array($this->m_aMailErrors) && count($this->m_aMailErrors) > 0)
-			{
+		} else {
+			if (is_array($this->m_aMailErrors) && count($this->m_aMailErrors) > 0) {
 				$sError = implode(', ', $this->m_aMailErrors);
-			}
-			else
-			{
+			} else {
 				$sError = 'Unknown reason';
 			}
+
 			return 'Notification was not sent: '.$sError;
 		}
 	}
+
+	/**
+	 * @param \DBObject $oObject
+	 * @param string $sHeaderName {@see \ActionEmail::ENUM_HEADER_NAME_REFERENCES}, {@see \ActionEmail::ENUM_HEADER_NAME_MESSAGE_ID}
+	 *
+	 * @return string The formatted identifier for $sHeaderName based on $oObject
+	 * @throws \Exception
+	 * @since 3.0.1 N°4849
+	 */
+	protected function GenerateIdentifierForHeaders(DBObject $oObject, string $sHeaderName): string
+	{
+		$sObjClass = get_class($oObject);
+		$sObjId = $oObject->GetKey();
+
+		switch ($sHeaderName) {
+			case static::ENUM_HEADER_NAME_MESSAGE_ID:
+			case static::ENUM_HEADER_NAME_REFERENCES:
+				// Prefix
+				$sPrefix = sprintf('iTop_%s_%d', $sObjClass, $sObjId);
+				if ($sHeaderName === static::ENUM_HEADER_NAME_MESSAGE_ID) {
+					$sPrefix .= sprintf('_%f', microtime(true /* get as float*/));
+				}
+				// Suffix
+				$sSuffix = sprintf('@%s.openitop.org', MetaModel::GetEnvironmentId());
+				// Identifier
+				$sIdentifier = $sPrefix.$sSuffix;
+				if ($sHeaderName === static::ENUM_HEADER_NAME_REFERENCES) {
+					$sIdentifier = "<$sIdentifier>";
+				}
+				break;
+
+			default:
+				$sErrorMessage = sprinf('%s: Could not generate identifier for header "%s", only %s are supported', static::class, $sHeaderName, implode(' / ', [static::ENUM_HEADER_NAME_MESSAGE_ID, static::ENUM_HEADER_NAME_REFERENCES]));
+				IssueLog::Error($sErrorMessage, LogChannels::ACTION, [
+					'Object' => $sObjClass.'::'.$sObjId.' ('.$oObject->Get('friendlyname').')',
+					'Action' => get_class($this).'::'.$this->GetKey().' ('.$this->Get('friendlyname').')',
```

Comment:
```suggestion
					'Object' => $sObjClass.'::'.$sObjId.' ('.$oObject->GetRawName().')',
					'Action' => get_class($this).'::'.$this->GetKey().' ('.$this->GetRawName().')',
```

Also, the `DBObject` class should have a `__toString`-like method that does this in one go. Unfortunately, the existing one returns html representation. (a `ToRawString` or something)

Reply:
Exactly my thought, we repeat this pattern on several occasion and it would be great to have it, but I'm didn't find a good name for it yet.

Reply:
Yeah, looking at the other existing methods, `ToRawString` or `GetAsRawString` would suffice. I also had a look and maybe we can just adapt the existing `__toString` method:
```php
	/**
	 * String representation of the object
	 *
	 * Only a few meaningful information will be returned.
	 * This representation is for debugging purposes, and is subject to change.
	 *
	 * @param bool $bHTML The returned string is raw HTML
	 *
	 * @return string
	 * @throws \ArchivedObjectException
	 * @throws \CoreException
	 * @throws \DictExceptionMissingString
	 */
	public function __toString($bHTML = true)
	{
        $sRet = '';
        $sClass = static::GetClassName(get_class($this));
        $iPKey = $this->GetKey();
        if ($bHTML) {
	        $sRootClass = MetaModel::GetRootClass($sClass);
	        $sFriendlyName = $this->GetAsHTML('friendlyname');
	        $sRet .= "<b title=\"$sRootClass\">$sClass</b>::$iPKey ($sFriendlyName)<br/>\n";
        } else {
        	$sFriendlyName = $this->GetRawName();
        	$sRet .= sprintf('%s::%d (%s)', $sClass, $iPKey, $sFriendlyName);
        }
        return $sRet;
	}
```